js/node-manager/src/client/agents/node-manager.tsx

```diff
@@ -0,0 +1,426 @@
+        trackedAddresses: ImmutableContainer(new Set()),
+        connections: ImmutableContainer(new Map()),
+        nodeDetailsProviders: ImmutableContainer(new Map()),
+        offsets: null as OffsetInfo | null,
+      })
+
+      const getAllTrackedAddrs = ImmutableLazyCalc(data.trackedAddresses, (trackedAddrs) =>
+        Array.from(trackedAddrs),
+      )
+
+      const getReachableUiNodes = ImmutableLazyCalc(data.nodeDetailsProviders, (providers) =>
+        Array.from(providers.values())
+          .map((provider) => provider.api.getAsReachableNodeUi())
+          .filter((info): info is ReachableNodeUi => info !== null),
+      )
+
+      const getConnectedNodes = ImmutableLazyCalc(data.connections, (connections) =>
+        Array.from(connections.entries())
+          .map(([addr, connection]) => {
+            if (!E.isRight(connection)) return null
+            return { addr, peer: connection.right.peer }
+          })
+          .filter((x): x is { addr: string; peer: string } => x !== null),
+      )
+
+      const getNodesAsUiNode = ImmutableLazyCalc(data.trackedAddresses, (addresses) =>
+        ImmutableLazyCalc(data.connections, (connections) =>
+          ImmutableLazyCalc(data.nodeDetailsProviders, (infoproviders) => {
+            const alltracked = Array.from(addresses)
+
+            return alltracked.map(
+              (addr): UiNode =>
+                intoUiNode(
+                  addr,
+                  connections.get(addr) || null,
+                  infoproviders.get(addr)?.api.get() || null,
+                ),
+            )
+          }),
+        ),
+      )
+
+      const unsubs = [
+        channels.onNodeDisconnect.sub((addr) => {
+          data.connections.mutate((connection) => {
+            console.log('disconnected from:', addr)
+            connection.delete(addr)
+          })
+          regulateRemovals(data, channels.change.emit)
+        }),
+        channels.onNodeInfoChange.sub((addr) => {
+          // swap node detail providers - triggering cache-invalidation to `getNodesAsUiNode`
+          data.nodeDetailsProviders.mutate((x) => x)
+          // refresh offsets everytime there's a change in NodeInfoChange
+          data.offsets = OffsetInfo.of(getReachableUiNodes())
+        }),
+      ]
+
+      const purge = () => {
+        console.log('node manager purged')
+        data.trackedAddresses.mutate((x) => x.clear())
+        regulateRemovals(data, channels.change.emit)
+      }
+
+      onDestroy(() => {
+        unsubs.map((unsub) => unsub())
+      })
+
+      // Recurring task
+      ;(async () => {
+        while (!isDestroyed()) {
+          // const getTimeoutSec =
+          //   (store.key === StoreStateKey.Loaded && store.data.preferences.nodeTimeout) ||
+          //   DEFAULT_TIMEOUT_SEC
+
+          if (allowedToWorkRef.get()) {
+            const getTimeoutSec = DEFAULT_TIMEOUT_SEC
+            regulateRemovals(data, channels.change.emit)
```

Comment:
This is a bit undocumented, but `regulateRemovals` is intended to be able to be run concurrently. This also benefits from it being a sync function that will not overlap each other due to how JS [event loop](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Event_loop#event_loop) is specced

Although there must only be one instance of running `regulateConnections` at a time.
I've put an extra note on the function's comment.

In other scenarios, it might make more sense that these exposed methods pipe parameter objects into a queue that will processed in a critical-section-locked processor similar to the async while in the center of the node-manager agent. My main reason for not implementing this is because, without being able to differentiate `mut` and not (like in rust), it will require quite a hassle to implement a separate pair of push-only API and a spin-pop, and it will make it less idiomatic toward JS devs.

Reply:
To be clear: my point was generic, and in this case it would only be actually problematic if `regulateRemovals` was an async function (which is as you rightly point out the only way to obtain concurrency in JS).